File: cpp/middle/sum_root_to_leaf_numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        if(!root)
            return 0;
        string path_sum;
        int sum = 0;
        getSum(root, sum, path_sum);
        return sum;
    }
    
    void getSum(TreeNode* root, int& sum, string path)
    {
        if(!root)
            return;
        path += to_string(root->val);
        if(root->left == NULL && root->right == NULL)
        {
            sum += stoi(path);
            return;
        }
        getSum(root->left, sum, path);
        getSum(root->right, sum, path);
    }
};


Conclusion:
比较简单的一道题，dfs preorder就可以处理这个问题。

Time Complexity: O(n)
Space Complexity: O(1)

File: python/middle/sum_root_to_leaf_numbers.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.path_sum = 0
        
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        path = ""
        self.getSum(root, path)
        return self.path_sum
    
    def getSum(self, root, path):
        if not root:
            return
        path += str(root.val)
        if root.left is None and root.right is None:
            self.path_sum += int(path)
        self.getSum(root.left, path)
        self.getSum(root.right, path)
        return
